# Profitable App Profiles for the App Store & Google Play Markets

- `What the project is about:` Pretending to work as a data analyst for a company that builds Android and iOS apps, which could be found in the App Store and Google Play. At this company, we only build apps that are free to download, which means our main source of revenue consists of in-app ads. This means that the number of users of our apps determines our revenue for any given app. 

- `Goal of this project:` Want to analyze data to help our developers understand what type of apps are likely to attract more users - the more users who see and engage with the ads, the better!


## Opening & Exploring Datasets

As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play

Collecting data for over 4 million apps requires a significant amount oif time and money, so we'll try to analyze a sample of the data instead. 

`A dataset` containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018. Can be found [Link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv)

`A dataset` containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017. Can be found [Link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv)

In [3]:
from csv import reader

# Importing AppleStore data 

open_file = open('AppleStore.csv')
read_file = reader(open_file)
ios = list(read_file)
ios_header = ios[0] #Isolating the Header Row
ios = ios[1:] #Removing header row from dataset

# Importing GooglePlay data

open_file = open('googleplaystore.csv')
read_file = reader(open_file)
android = list(read_file)
android_header = android[0]  #Isolating the Header Row
android = android[1:] #Removing header row from dataset

We'll start by opening and exploring these two data sets. To make them easier to explore, we created a function named `explore_data()` that you can repeatedly use to print rows in a readable way.

In [4]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [5]:
#Printing first few rows of App Store dataset

print(ios_header)
print('\n')
explore_data(ios, 0, 4, True) #Set to True to get number of rows and columns


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7197
Number of columns: 16


For the `App Store Dataset`, we found that there are 7,197 rows and 16 columns. Columns that could help us with our analysis include: `track_name`, `currency`, `price`, `rating_count_tot`, and `prime_genre`. Some column names are not descriptive enough, however the documentation can be found [Here](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home)

In [6]:
#Printing first few rows of Android Dataset

print(android_header)
print('\n')
explore_data(android, 0, 4, True) #Set to True to get number of rows and columns

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10841
Number of columns: 13


As for the `Android dataset`, we found that there are 10,841 rows and 13 columns within the dataset. Columns that could help us with our analysis include: `App`, `Category`, `Rating`, `Reviews`, `Installs`, `Price`, and `Genres`. 

## Data Cleaning

Recall that at our company, we only build apps that are free to download and install, and we design them for an *English-speaking* audience. This means that we'll need to do the following:

- Remove non-English apps
- Remove apps that are not Free

## Removing Row with missing value

The Google Play dataset has a dedicated [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), and we can see that [one of the discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) describes an error for row 10472.

In [7]:
print(android_header) # List of all column headers
print('\n')
print(android[10472]) # Error row
print('\n')
print(android[1]) # Correct row to compare with error row

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


When lining up the error row and correct row, we see that row 10472 is missing a value in the `Category` column. Because of this, we want to remove the row using the `del` [statement](https://docs.python.org/3/reference/simple_stmts.html?highlight=del#the-del-statement)

In [8]:
print(len(android)) # check length prior
del android[10472]  # don't run this more than once
print(len(android)) # make sure only one row is deleted

10841
10840


## Removing Duplicate Entries

If we explore the Google Play dataset long enough or look at the [discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion) section, we'll notice some apps have duplicate entries. For instance, Instagram has four entries:

In [9]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In total, there are 1,181 cases where an app occurs more than once. 

In [10]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:15])

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


We don't want to count certain apps more than once when we analyze data, so we need to remove the duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, but we could probably find a better way.

If you examine the rows we printed for the Instagram app, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show the data was collected at different times. We can use this information to build a criterion for removing the duplicates. The higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, we'll only keep the row with the highest number of reviews and remove the other entries for any given app.

To remove the duplicates, we do the following:

- Create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app.
- Use the information stored in the dictionary and create a new dataset, which will have only one entry per app (and for each app, we'll only select the entry with the highest number of reviews)


In [11]:
reviews_max = {}  # creating empty dictionary

for app in android:  # loop thru Google Play data set
    name = app[0]    # app name
    n_reviews = float(app[3])   # converting num of reviews to float
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

In previous code chunk above, we found that there were 1,181 duplicate apps, which means we should be able to find the true number of unique apps by subtracting 1,181 from the length of the Google Play dataset. 

In [12]:
print('Expected length:', len(android)-1181) 
print('Actual length:', len(reviews_max))  # inspecting dictionary to make sure everything went as expected

Expected length: 9659
Actual length: 9659


Using the dictionary we created above to remove the duplicate rows:

- Start by creating two empty lists: `android_clean` (which will store our new cleaned data set) and `already_added` (which will just store app names).
- Loop through the Google Play dataset (don't include the header row), and for each iteration, do the following:
    - Assign the app name to a variable named `name`.
    - Convert the number of reviews to `float`, and assign it to a variable named `n_reviews`.
- If `n_reviews` is the same as the number of maximum reviews of the app `name` (the number can be found in the `reviews_max` dictionary) and `name` is not already in the list `already_added`:
    - Append the entire row to the `android_clean` list (which will eventually be a list of lists and store our cleaned dataset).
    - Append the name of the app `name` to the `already_added` list — this helps us to keep track of apps that we already added.

In [13]:
android_clean = [] # empty list which will store new cleaned dataset
already_added = [] # will just store app names

for app in android:
    name = app[0]    
    n_reviews = float(app[3])
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)  # will eventually be a list of lists and store our cleaned dataset
        already_added.append(name) # helps us keep track of apps that we already added


In [14]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


Exploring the `android_clean` dataset to ensure everything went as expected. The dataset should have 9,659 rows, and it does. 

## Removing Non-English Apps

## Part One

Remember we use English for the apps we develop at our company, and we'd like to analyze only the apps that are designed for an English-speaking audience. However, if we explore the data long enough, we'll find that both datasets have apps with names that suggest they are not designed for an English-speaking audience

In [15]:
print(ios[813][1])
print(ios[6731][1])
print('/n')
print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
/n
中国語 AQリスニング
لعبة تقدر تربح DZ


We're not interested in keeping these kind of apps, so we'll remove them. One way to go about this is to remove each app whose name contains a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, *, /, etc.).

All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.

We built this function below, and we use the built-in `ord()` function to find out the corresponding encoding number of each character.


In [16]:
def english_or_nah(a_string):
    for character in a_string:   # iterate over input string
        if ord(character) > 127: 
            return False         # if character is greater than 127, app name is probably non-English
    return True                  # app name is probably English


print(english_or_nah('Instagram'))
print(english_or_nah('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english_or_nah('Docs To Go™ Free Office Suite'))
print(english_or_nah('Instachat 😜'))

True
False
False
False


The function seems fine when checking the first two app names, however struggles with the third and fourth app names (although they are English app names). This is because emojis and characters like ™ fall outside the ASCII range and have corresponding numbers over 127

In [17]:
print(ord("™"))
print(ord("😜"))

8482
128540


## Part Two

If we're going to use the function we've created, we'll lose useful data since many English apps will be incorrectly labeled as non-English. To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This means all English apps with up to three emoji or other special characters will still be labeled as English.

In [18]:
def english_or_nah(a_string):
    
    outside_ascii = 0
    
    for character in a_string:   
        if ord(character) > 127: 
            outside_ascii += 1
            
    if outside_ascii > 3:
        return False
    else: 
        return True
    
    
print(english_or_nah('Docs To Go™ Free Office Suite'))
print(english_or_nah('Instachat 😜'))
print(english_or_nah('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


 Our filter function is still not perfect, but it should be fairly effective.

In [19]:
ios_english = []
android_english = []

for app in ios:
    name = app[1]
    if english_or_nah(name):
        ios_english.append(app)

for app in android_clean:
    name = app[0]
    if english_or_nah(name):
        android_english.append(app)
        
explore_data(ios_english, 0, 3, True)
print('\n')
explore_data(android_english, 0, 3, True)


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 6183
Number of columns: 16


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Vari

Exploring the new datasets, we see that we now have 6183 iOS apps and 9614 Android apps.

## Isolating the Free Apps

In [20]:
ios_free = []
android_free = []

for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_free.append(app)
        
for app in android_english:
    price = app[7]
    if price == '0':
        android_free.append(app)

print(len(ios_free))
print(len(android_free))

3222
8864


After isolating the free apps, we get 3,222 iOS apps and 8,864 Android apps remaining.

## Most Common Apps by Genre

## Part One

We want to find an app profile that fits both the App Store and Google Play because we want to attract more users as the number of people using our apps affect our revenue. To minimize risk and overhead, our validation strategy for an app idea has three steps:

1. Build a minimal Android version of the app, and add it to Google Play
2. If the app has a good response from users, we develop it further
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful in both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.


We will begin the analysis by determining the most common genres for each market. For this, we'll need to build frequency tables using the `prime_genre` column from the App Store dataset, and the `Genres` and `Category` columns from the Google Play dataset

## Part Two

We'll build two functions we can use to analyze the frequency tables:

- One function to generate frequency tables that show percentages
- Another function we can use to display the percentages in a descending order

In [38]:
def freq_table(dataset, index):  
    table_freq = {}
    count = 0
    for row in dataset:
        count += 1
        value = row[index]
        if value in table_freq:
            table_freq[value] += 1
        else:
            table_freq[value] = 1
            
    percentages = {}
    for key in table_freq: 
        percentage = (table_freq[key] / count) * 100
        percentages[key] = percentage
    return percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

## Part Three

In [43]:
display_table(ios_free, 11)  # Freq table for prime_genre column of the App Store

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Analyzing the frequency table that we generated for the `prime_genre` column of the App Store dataset, we see that an overwhelming 58.2% of English apps in the store are of the Games genre. The next most common genre being Entertainment, and a distant 7.9% of English apps in the dataset. The least common genre happened to be Catalogs, with only a 0.12% share of the apps in the dataset.

My general impression is that most of the apps in this App Store dataset are designed for entertainment (games, photo & video, social networking, sports, music, etc). Those genres combined are responsible for over 70% of the apps in the App Store. It's hard to recommend an app profile for the App Store market based on this frequency table alone. A large number of apps for a particular genre does not imply that apps of that genre generally have a large number of users. 

In [44]:
display_table(android_free, 1)   # Freq table for Category column of the Google Play store 

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

Analyzing the `Category` column of the Google Play store, we see that the most common genre is the family genre which occupies approximately 19% of English apps in the store. The next two most common genres are Games and Tools, which take up approximately 9.7% and 8.5% of English apps. Compared to the App Store market, there isn't a genre that dominates the market as much as Games do in the App Store. There seems to be more parity in the Google Play store relative to the App Store. Based on our findings, we see more balance, as it does not lean too heavily in one direction.    

In [46]:
display_table(android_free, -4)   # Freq table for Genre column of the Google Play store 

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

Analyzing the `Genre` column of the Google Play store, we see that the Tools genre is the most common, followed by Entertainment and Education. This frequency table shows a lot more genres than the previous table where we indexed `Category`. 

## Most Popular Apps by Genre on the App Store

The frequency tables we analyzed on the previous screen showed us that apps designed for fun dominate the App Store, while Google Play shows a more balanced landscape of both practical and fun apps. Now, we'd like to determine the kind of apps with the most users.

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the `Installs` column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the `rating_count_tot` app.

To calculate the average number of user ratings for each game, we'll use a for loop inside of another loop. 

In [48]:
ios_freq_table = freq_table(ios_free, 11)  # Freq table for prime_genre column of the App Store

for genre in ios_freq_table:
    total = 0
    len_genre = 0
    for app in ios_free:
        genre_app = app[11]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_ratings = total / len_genre
    print(genre, ':', avg_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


Analyzing the results, we find that the Navigation genre has the highest average number of user ratings at approximately 86,090 reviews. The next highest average number of user ratings for a genre are Reference and Social Networking, at 74,942 and 71,548, respectively. 

## Most Popular Apps by Genre on Google Play

We have data about the number of installs for the Google Play market, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [49]:
display_table(android_free, 5)   # The Installs columns

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


A weakness of this specific dataset is that we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to find out which app genres attract the most users

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on. To perform computations, however, we'll need to convert each install number from a string to a float. This means we need to remove the commas and the plus characters, or the conversion will fail and cause an error.

In [52]:
android_freq = freq_table(android_free, 1)

for category in android_freq:  # Loop over unique genres of Google Play dataset
    total = 0    # variable will store sum of installs specific to each genre
    len_category = 0     # will store the number of apps specific to each genre
    for app in android_free:       # loop over google play dataset
        category_app = app[1]         # save app genre to var
        if category_app == category:            
            n_installs = app[5]           # save number of installs
            n_installs = n_installs.replace(',', '')  # remove + or , and convert str to float
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)         
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

Analyzing the results, we find that the most popular apps on the Google Play store are that of the Communication genre, with an average of 38,456,119 installs. The next most installed apps by genre are that of Social and Video_Players, with averages of 23,253,652 and 24,727,872 installs, respectively. 

## Conclusion

This project was centered around building and developing free English-langugage apps and how we could increase the amount of people that use our apps as that determines the amount of revenue for a given app. With that being said, we went through two datasets, with one consisting data from the App Store and the other from the Google Play Store.

We found the most popular genres of both the App Store and Google Play Store. However, those popular genres may be dominated by a singular or multiple giants, which in turn may have skewed how popular some of the genres may actually be. As a result of that, it would be hard to profit in a genre that is being dominated by the bigger, well-known apps. So through analyzing the results, I would recommend an app of the Education genre as it could potentially show potential for being profitable on both the App Store and Google Play. Especially with the on going pandemic, and virtual schooling being so